### ISSUE TITLE
Should `std::error_code::message` respect the locale set by the user?

### ISSUE THREAD

**Issue by solarispika** [microsoft/STL#4711]
In our program, we use `std::setlocale(LC_ALL, ".UTF8");` to support UTF-8 encoding, as noted in the [Microsoft documentation](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170#utf-8-support). However, when certain C++ STL functions fail, their `message()` outputs are not in UTF-8.

After investigating the source code, I found that the error messages are always formatted according to the system's default locale, as shown in this STL implementation:
https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/src/syserror_import_lib.cpp#L38-L54

While it would be acceptable if the system locale used UTF-8 as the codepage, allowing the messages to be correctly printed to log files, this is not the case. There is a "Beta: Use Unicode UTF-8 for worldwide language support" setting in the Region settings, which does the job correctly, but toggling that setting requires a reboot, which is not acceptable for our requirements.

## Question

Should the `std::error_code::message` function respect the locale set by the user (e.g., via `std::setlocale`), or should it continue to use the system's default locale? If the former, is there a plan to implement this behavior in the Microsoft STL implementation?

## Related Information

- I learned that the locale is by default set across the whole program, unless specified by [configthreadlocale](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/configthreadlocale?view=msvc-170), from the Microsoft documentation.
- The issue of system_error not honoring the current thread's locale is mentioned in [this issue](https://github.com/microsoft/STL/issues/2882).

Please let me know if you need any additional information or clarification.

**Comment by jovibor** [comment]
This is basically a duplicate of the #3254.
Also see the #3260.

**Comment by miyanyan** [comment]
I have the same question, when I get filesystem_error, it's message often uses the system locale, and I have to change the message to utf8

**Comment by StephanTLavavej** [comment]
@jovibor Yep, agreed. Marking as "bug" but not resolving as a duplicate, until we review the pending PR.

**Comment by vitaut** [comment]
The standard is unclear what encoding `error_category::what` should use - I've filed an LWG issue about it: https://cplusplus.github.io/LWG/issue4156. The current proposed resolution there is a bit of a strawman and likely to change. I would recommend waiting until this LWG issue is resolved.

**Comment by vitaut** [comment]
FYI SG16 just voted in favor of specifying that the string returned by `error_category::what()` is in the C locale (aka execution) encoding as part of [P3395](https://isocpp.org/files/papers/P3395R1.html). The `formatter` specialization for `error_code` was made more conservative and left the transcoding decision to the implementation if the literal encoding is not UTF-8. Let me know if you have any concerns regarding these decisions.

**Comment by cpplearner** [comment]
The standard says that the encoding of the execution character set is locale-specific ([\[character.seq.general\]/(1.2)](https://eel.is/c++draft/character.seq.general#1.2)):

> The encodings of the execution character sets and the sets of additional elements (if any) are locale-specific.

It's unclear which locale should be used. Possible choices are

- The "C" locale with UTF-8 encoding (`setlocale("C.UTF-8", LC_CTYPE)` if supported)
- The "C" locale (`setlocale("C", LC_CTYPE)`)
- The currently-installed global locale, whatever it is (`setlocale(nullptr, LC_CTYPE)`)
- Something else (like `setlocale("", LC_CTYPE)`)
